### PR TITLE
ref(releases): Add commit/files pagination

### PR DESCRIPTION
**Pull request by priscilawebdev**
**Description:**

This PR adds commits/files changed pagination

closes: https://app.asana.com/0/1158284503473033/1177990501294411

it looks like:

commits
image

files changed

image

#sync-getsentry

### CODE REVIEW

File: src/sentry/static/sentry/app/views/releasesV2/detail/utils.tsx [hunk 1]
```diff
@@ -61,3 +65,35 @@ export function getCommitsByRepository(commitList: Commit[]): CommitsByRepositor
...
+
+/**
+ * Get request query according to the url params and active repository
+ */
+
+type GetQueryProps = {
+  location: Location;
+  perPage?: number;
+  activeRepository?: Repository;
+};
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/releasesV2/detail/utils.tsx:77]:
```suggestion
};

```

File: src/sentry/api/endpoints/project_release_repositories.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        """
+
+        try:
+            release = Release.objects.get(
+                organization_id=project.organization_id, projects=project, version=version
+            )
+        except Release.DoesNotExist:
+            raise ResourceDoesNotExist
+
+        release_commits = ReleaseCommit.objects.filter(release=release).select_related("commit")
```
Comment (markstory) [inline src/sentry/api/endpoints/project_release_repositories.py:38]: If you only need the repositories, you could use `.select_related("commit__repository")` to get everything in fewer queries.

Comment (priscilawebdev) [reply src/sentry/api/endpoints/project_release_repositories.py:38]: Thank you for pointing this out, but unfortunately, it doesn't seem to have an association.

Comment (untitaker) [reply src/sentry/api/endpoints/project_release_repositories.py:38]: I'd try `commit__repository_id`

Comment (untitaker) [reply src/sentry/api/endpoints/project_release_repositories.py:38]: nvm that doesn't work because it would not be an extra join

**Review by untitaker — APPROVED** [review]
only reviewed python files

**Comment by matejminar** [comment]
Do we need to reset `cursor` in the URL when switching between tabs (Overivew, Commits, Files Changed, Artifacts)?

**Review by matejminar — COMMENTED** [review]
~~Should you reset the cursor in the URL when switching between tabs (Overview, Commits, Files Changed, Artifacts)?~~

EDIT: Sorry, it already works like that.

**Comment by priscilawebdev** [comment]
@matejminar yes, imo it doesn't make sense to keep the cursor that belongs to some specific tab in the URL when switching tabs.

**Comment by matejminar** [comment]
image

Why do we show 2 repositories when there is only one involved in this release?
I'd expect to see only one in the dropdown, in fact, when there is only one repository involved (which should be the majority of use cases) I'd hide the dropdown altogether.

Also, does the `All Repositories` option make sense? It's imho kind of confusing when you have a different number of files/commits per each repo on every page. I'd probably just select the first option by default and remove the `All Repositories` option.

**Comment by priscilawebdev** [comment]
@matejminar thank you for the feedback :) I've applied some of your suggestions. There is one special case when we display repositories that are not related to any of the changed files. If I remember well, this happens when we use a 'skip" flag and don't ingest the commit's files..maybe @markstory could explain it better to you .... what I can say is that it will probably happen more often for us internally because it's common to see users using this flag.

**Comment by markstory** [comment]
> If I remember well, this happens when we use a 'skip" flag and don't ingest the commit's files..maybe @markstory could explain it better .... what I can say is that it will probably happen more often for us internally because it's not common to see users using this flag.

The commits with no file contents in our account show up because of a combination of how we use sentry, and how we do deploys.

* When handling push webhooks from repository integrations we ignore any commits that contain `#skipsentry` in the commit message. This lets users exclude commits from suspect commit tracing if they need to. We primarily use it for the automated commits that sentrybot makes to getsentry.
* Then when we do a deploy we collect the shas in the release for both sentry and getsentry. Freight does not exclude commits with the `#skipsentry` message though. 
* The collected commits are included as 'release commits' for the release. Because we don't have any commits records for these commits (because we skipped them) the commits are created during deploy.
* Because all we have is the sha an empty commit with no message and no changed files is created in our database.

File: src/sentry/static/sentry/app/views/releasesV2/detail/commits.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    const {commits, commitsPageLinks} = this.state;
+    const {activeRepository} = this.props;
+
+    if (!commits.length) {
+      return (
+        <EmptyState>
+          {!activeRepository
+            ? t('There are no commits associated with this release.')
+            : t(
+                'There are no commits associated with this release in the repository %s.',
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/releasesV2/detail/commits.tsx:77]:
```suggestion
                'There are no commits associated with this release in the %s repository.',
```

File: src/sentry/static/sentry/app/views/releasesV2/detail/filesChanged.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    const {fileList, fileListPageLinks} = this.state;
+    const {activeRepository} = this.props;
+
+    if (!fileList.length) {
+      return (
+        <EmptyState>
+          {!activeRepository
+            ? t('There are no changed files associated with this release.')
+            : t(
+                'There are no changed files associated with this release in the repository %s.',
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/releasesV2/detail/filesChanged.tsx:76]:
```suggestion
                'There are no changed files associated with this release in the %s repository.',
```

File: src/sentry/static/sentry/app/views/releasesV2/detail/withRepositories/index.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      this.setState({isLoading: true});
+      try {
+        const repositories = await this.props.api.requestPromise(this.getEndpoint());
+        this.setState({repositories, isLoading: false});
+      } catch (error) {
+        Sentry.captureException(error);
+        addErrorMessage(
+          t(
+            `An error occured while trying to fetch the repositories of the release: ${release}`
+          )
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/releasesV2/detail/withRepositories/index.tsx:89]:
```suggestion
          t(
            'An error occured while trying to fetch the repositories of the release: %s',
            release
          )
```

File: src/sentry/static/sentry/app/views/releasesV2/detail/repositorySwitcher.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    > *:nth-child(2) {
+      right: auto;
+      width: auto;
+      ${p => p.minMenuWidth && `min-width: calc(${p.minMenuWidth}px + 10px);`}
+      border-radius: ${p => p.theme.borderRadius};
+      border-top-left-radius: 0px;
+      border: 1px solid ${p => p.theme.button.default.border};
+      top: calc(100% - 1px);
+    }
+  `;
```
Comment (matejminar) [inline src/sentry/static/sentry/app/views/releasesV2/detail/repositorySwitcher.tsx:98]: What are these custom styles? Should they be a part of DropdownControl?

**Review by matejminar — COMMENTED** [review]
It frustrates me a little that we return a different amount of files on every page:
Kapture 2020-09-23 at 09 53 24
Is there anything we can do about it?

File: src/sentry/static/sentry/app/views/releasesV2/detail/repositorySwitcher.tsx (hunk 6 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/releasesV2/detail/repositorySwitcher.tsx:98]: I correct one problem that I noticed in the dropDownControl.... I plan to move this code in another PR

**Comment by priscilawebdev** [comment]
@matejminar I talked to backend about moving the logic of the function "getFilesByRepository" to the backend in order to get the 20 files and we conclude that there are two options, one being expensive and the second unstable.

First option:
Fetch everything from the database, collapsing some results according to the filename (as we currently do in the frontend)

-> this can be very expensive and it doesn't make much sense as we are using pagination

Second option:

We could send multiple requests with an offset  to get more items until we get a total of 20 items, but this can be very unstable. It's not recommended.


In the end we came to the conclusion that the most viable would be to leave the UI as it is 😕

**Comment by matejminar** [comment]
Can we at least increase the limit of items returned? So that you see at least ~20 (or even 50?) at a time?

**Review by matejminar — APPROVED** [review]

**Pull request merged**